File: backend/worker/handlers/concat_vertical_timeline.py
import asyncio
import os
import tempfile

from worker.handlers.base import BaseHandler


class ConcatVerticalTimelineHandler(BaseHandler):
    async def execute(self, node_config, input_paths, output_path):
        first_video = input_paths["video_first"]
        second_video = input_paths["video_second"]
        top_image = input_paths.get("image_top")
        bottom_image = input_paths.get("image_bottom")

        pane_width = int(node_config.get("pane_width", 640) or 640)
        pane_height = int(node_config.get("pane_height", 360) or 360)
        background_color = str(node_config.get("background_color", "black") or "black")

        generated_top_image = None
        generated_bottom_image = None
        if not top_image:
            generated_top_image = await self._extract_default_frame(
                video_path=first_video,
                prefer_from_end=True,
            )
            top_image = generated_top_image
        if not bottom_image:
            generated_bottom_image = await self._extract_default_frame(
                video_path=second_video,
                prefer_from_end=False,
            )
            bottom_image = generated_bottom_image

        first_segment = await self._render_segment(
            active_video=first_video,
            static_image=bottom_image,
            active_position="top",
            pane_width=pane_width,
            pane_height=pane_height,
            background_color=background_color,
        )
        second_segment = await self._render_segment(
            active_video=second_video,
            static_image=top_image,
            active_position="bottom",
            pane_width=pane_width,
            pane_height=pane_height,
            background_color=background_color,
        )

        try:
            filter_complex = "[0:v][0:a][1:v][1:a]concat=n=2:v=1:a=1[v][a]"
            args = [
                "-i", first_segment,
                "-i", second_segment,
                "-filter_complex", filter_complex,
                "-map", "[v]",
                "-map", "[a]",
                *self.build_video_encode_args("libx264", preset="fast", crf=23),
                "-c:a", "aac",
                output_path,
            ]
            await self.run_ffmpeg(args)
        finally:
            self._safe_unlink(first_segment)
            self._safe_unlink(second_segment)
            if generated_top_image:
                self._safe_unlink(generated_top_image)
            if generated_bottom_image:
                self._safe_unlink(generated_bottom_image)

    async def _render_segment(
        self,
        *,
        active_video: str,
        static_image: str,
        active_position: str,
        pane_width: int,
        pane_height: int,
        background_color: str,
    ) -> str:
        probe = await self.run_ffprobe(active_video)
        duration = float(probe.get("format", {}).get("duration", 0) or 0)
        if duration <= 0:
            duration = 5.0

        has_audio = any(stream.get("codec_type") == "audio" for stream in probe.get("streams", []))
        output_file = tempfile.NamedTemporaryFile(suffix=".mp4", delete=False)
        output_file.close()

        if active_position == "top":
            top_source = "[0:v]"
            bottom_source = "[1:v]"
        else:
            top_source = "[1:v]"
            bottom_source = "[0:v]"

        filter_complex = (
            f"{top_source}scale={pane_width}:{pane_height}:force_original_aspect_ratio=decrease,"
            f"pad={pane_width}:{pane_height}:(ow-iw)/2:(oh-ih)/2:color={background_color},fps=30[top];"
            f"{bottom_source}scale={pane_width}:{pane_height}:force_original_aspect_ratio=decrease,"
            f"pad={pane_width}:{pane_height}:(ow-iw)/2:(oh-ih)/2:color={background_color},fps=30[bottom];"
            "[top][bottom]vstack=inputs=2[v]"
        )

        args = [
            "-i", active_video,
            "-loop", "1",
            "-t", f"{duration:.3f}",
            "-i", static_image,
        ]

        if not has_audio:
            args.extend([
                "-f", "lavfi",
                "-t", f"{duration:.3f}",
                "-i", "anullsrc=r=44100:cl=stereo",
            ])

        args.extend([
            "-filter_complex", filter_complex,
            "-map", "[v]",
        ])

        if has_audio:
            args.extend(["-map", "0:a:0"])
        else:
            args.extend(["-map", "2:a:0"])

        args.extend([
            *self.build_video_encode_args("libx264", preset="fast", crf=23),
            "-c:a", "aac",
            "-pix_fmt", "yuv420p",
            "-shortest",
            output_file.name,
        ])

        await self.run_ffmpeg(args)
        return output_file.name

    async def _extract_default_frame(self, *, video_path: str, prefer_from_end: bool) -> str:
        frame_count = await self._count_video_frames(video_path)
        if frame_count <= 0:
            frame_count = 1

        target_offset = 15
        if prefer_from_end:
            frame_index = max(frame_count - target_offset, 0)
        else:
            frame_index = min(target_offset - 1, frame_count - 1)

        output_file = tempfile.NamedTemporaryFile(suffix=".png", delete=False)
        output_file.close()

        filter_expr = f"select=eq(n\\,{frame_index})"
        args = [
            "-i", video_path,
            "-vf", filter_expr,
            "-vsync", "vfr",
            "-frames:v", "1",
            output_file.name,
        ]
        await self.run_ffmpeg(args)
        return output_file.name

    async def _count_video_frames(self, video_path: str) -> int:
        cmd = [
            "ffprobe",
            "-v", "error",
            "-select_streams", "v:0",
            "-count_frames",
            "-show_entries", "stream=nb_read_frames",
            "-of", "default=nokey=1:noprint_wrappers=1",
            video_path,
        ]
        proc = await asyncio.create_subprocess_exec(
            *cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, _stderr = await proc.communicate()
        if proc.returncode != 0:
            return 0
        raw = stdout.decode("utf-8", errors="replace").strip()
        try:
            return int(raw)
        except ValueError:
            return 0

    def _safe_unlink(self, path: str) -> None:
        try:
            os.unlink(path)
        except FileNotFoundError:
            pass

File: internal/worker/handlers/concat_vertical_timeline.go
package handlers

import (
	"context"
	"errors"
	"os"
	"strconv"

	vpffmpeg "github.com/Ctwqk/videoprocess/internal/worker/ffmpeg"
)

type ConcatVerticalTimelineHandler struct {
	Runner vpffmpeg.Runner
}

func (h ConcatVerticalTimelineHandler) NodeType() string {
	return "concat_vertical_timeline"
}

func (h ConcatVerticalTimelineHandler) Execute(ctx context.Context, inputPaths map[string]string, outputPath string, config map[string]any) (map[string]any, error) {
	firstVideo := inputPaths["video_first"]
	if firstVideo == "" {
		return nil, errors.New("missing input path on video_first port")
	}
	secondVideo := inputPaths["video_second"]
	if secondVideo == "" {
		return nil, errors.New("missing input path on video_second port")
	}
	topImage := inputPaths["image_top"]
	bottomImage := inputPaths["image_bottom"]

	var generatedTop string
	var generatedBottom string
	var err error
	if topImage == "" {
		generatedTop, err = h.extractDefaultFrame(ctx, firstVideo, true)
		if err != nil {
			return nil, err
		}
		topImage = generatedTop
	}
	if bottomImage == "" {
		generatedBottom, err = h.extractDefaultFrame(ctx, secondVideo, false)
		if err != nil {
			safeRemove(generatedTop)
			return nil, err
		}
		bottomImage = generatedBottom
	}

	paneWidth := intValue(config["pane_width"], 640)
	paneHeight := intValue(config["pane_height"], 360)
	backgroundColor := stringValue(config["background_color"], "black")

	firstSegment, err := h.renderSegment(ctx, firstVideo, bottomImage, "top", paneWidth, paneHeight, backgroundColor)
	if err != nil {
		safeRemove(generatedTop)
		safeRemove(generatedBottom)
		return nil, err
	}
	secondSegment, err := h.renderSegment(ctx, secondVideo, topImage, "bottom", paneWidth, paneHeight, backgroundColor)
	if err != nil {
		safeRemove(firstSegment)
		safeRemove(generatedTop)
		safeRemove(generatedBottom)
		return nil, err
	}
	defer safeRemove(firstSegment)
	defer safeRemove(secondSegment)
	defer safeRemove(generatedTop)
	defer safeRemove(generatedBottom)

	args := []string{
		"-i", firstSegment,
		"-i", secondSegment,
		"-filter_complex", "[0:v][0:a][1:v][1:a]concat=n=2:v=1:a=1[v][a]",
		"-map", "[v]",
		"-map", "[a]",
	}
	args = append(args, intermediateVideoEncodeArgs("libx264")...)
	args = append(args, "-c:a", "aac", outputPath)
	if err := runFFmpeg(ctx, h.Runner, args); err != nil {
		return nil, err
	}
	return map[string]any{}, nil
}

func (h ConcatVerticalTimelineHandler) renderSegment(ctx context.Context, activeVideo string, staticImage string, activePosition string, paneWidth int, paneHeight int, backgroundColor string) (string, error) {
	probe, err := probePath(ctx, h.Runner, activeVideo)
	if err != nil {
		return "", err
	}
	if probe.Duration <= 0 {
		probe.Duration = 5
	}
	outputFile, err := os.CreateTemp("", "vp_vertical_segment_*.mp4")
	if err != nil {
		return "", err
	}
	outputPath := outputFile.Name()
	if err := outputFile.Close(); err != nil {
		safeRemove(outputPath)
		return "", err
	}

	args := VerticalTimelineSegmentArgs(activeVideo, staticImage, outputPath, activePosition, paneWidth, paneHeight, backgroundColor, probe)
	if err := runFFmpeg(ctx, h.Runner, args); err != nil {
		safeRemove(outputPath)
		return "", err
	}
	return outputPath, nil
}

func VerticalTimelineSegmentArgs(activeVideo string, staticImage string, outputPath string, activePosition string, paneWidth int, paneHeight int, backgroundColor string, probe probeSummary) []string {
	duration := probe.Duration
	if duration <= 0 {
		duration = 5
	}
	durationText := strconv.FormatFloat(duration, 'f', 3, 64)

	topSource := "[0:v]"
	bottomSource := "[1:v]"
	if activePosition != "top" {
		topSource = "[1:v]"
		bottomSource = "[0:v]"
	}

	filterComplex := topSource + scaleFilter(strconv.Itoa(paneWidth), strconv.Itoa(paneHeight), "decrease") +
		",pad=" + strconv.Itoa(paneWidth) + ":" + strconv.Itoa(paneHeight) + ":(ow-iw)/2:(oh-ih)/2:color=" + backgroundColor + ",setsar=1,fps=30[top];" +
		bottomSource + scaleFilter(strconv.Itoa(paneWidth), strconv.Itoa(paneHeight), "decrease") +
		",pad=" + strconv.Itoa(paneWidth) + ":" + strconv.Itoa(paneHeight) + ":(ow-iw)/2:(oh-ih)/2:color=" + backgroundColor + ",setsar=1,fps=30[bottom];" +
		"[top][bottom]vstack=inputs=2,setsar=1[v]"

	args := []string{
		"-i", activeVideo,
		"-loop", "1",
		"-t", durationText,
		"-i", staticImage,
	}
	if !probe.HasAudio {
		args = append(args,
			"-f", "lavfi",
			"-t", durationText,
			"-i", "anullsrc=r=48000:cl=stereo",
		)
	}
	args = append(args,
		"-filter_complex", filterComplex,
		"-map", "[v]",
	)
	if probe.HasAudio {
		args = append(args, "-map", "0:a:0")
	} else {
		args = append(args, "-map", "2:a:0")
	}
	args = append(args, intermediateVideoEncodeArgs("libx264")...)
	args = append(args,
		"-c:a", "aac",
		"-shortest",
		outputPath,
	)
	return args
}

func (h ConcatVerticalTimelineHandler) extractDefaultFrame(ctx context.Context, videoPath string, preferFromEnd bool) (string, error) {
	frameCount, err := h.countVideoFrames(ctx, videoPath)
	if err != nil {
		return "", err
	}
	frameIndex := defaultFrameIndex(frameCount, preferFromEnd)
	outputFile, err := os.CreateTemp("", "vp_vertical_frame_*.png")
	if err != nil {
		return "", err
	}
	outputPath := outputFile.Name()
	if err := outputFile.Close(); err != nil {
		safeRemove(outputPath)
		return "", err
	}
	args := []string{
		"-i", videoPath,
		"-vf", "select=eq(n\\," + strconv.Itoa(frameIndex) + ")",
		"-vsync", "vfr",
		"-frames:v", "1",
		outputPath,
	}
	if err := runFFmpeg(ctx, h.Runner, args); err != nil {
		safeRemove(outputPath)
		return "", err
	}
	return outputPath, nil
}

func (h ConcatVerticalTimelineHandler) countVideoFrames(ctx context.Context, videoPath string) (int, error) {
	runner := h.Runner
	if runner.Binary == "" && runner.ProbeBinary == "" {
		runner = vpffmpeg.NewRunner()
	}
	return runner.CountVideoFrames(ctx, videoPath)
}

func defaultFrameIndex(frameCount int, preferFromEnd bool) int {
	if frameCount <= 0 {
		frameCount = 1
	}
	targetOffset := 15
	if preferFromEnd {
		index := frameCount - targetOffset
		if index < 0 {
			return 0
		}
		return index
	}
	index := targetOffset - 1
	if index > frameCount-1 {
		return frameCount - 1
	}
	return index
}
